### ISSUE TITLE
Cannot build with .node files and Vercel adapter

### ISSUE THREAD

**Issue by Nick-Mazuk** [sveltejs/kit#1547]
**Describe the bug**

Cannot build Sveltekit while using Sharp and the Vercel adapter. Sharp works in development. I suspect this issue applies to all dependencies with `.node` files.

This seems unrelated to the ["error when including packages"](https://kit.svelte.dev/faq#packages) question in the FAQ or https://github.com/sveltejs/kit/issues/737

**Logs**
```bash
> Using @sveltejs/adapter-vercel
 > node_modules/.pnpm/sharp@0.28.2/node_modules/sharp/lib/output.js:4:22: error: No loader is configured for ".node" files: node_modules/.pnpm/sharp@0.28.2/node_modules/sharp/build/Release/sharp.node
    4 │ const sharp = require('../build/Release/sharp.node');
```

**To Reproduce**

1. Initialize new Sveltekit project
2. Install sharp `pnpm i sharp`
3. Create a new endpoint and import sharp
4. Add Vercel adapter
5. `pnpm build`

Reproduction repo: https://github.com/Nick-Mazuk/sveltekit-sharp

**Expected behavior**

Site would build without issues.

**Stacktraces**
If you have a stack trace to include, we recommend putting inside a `<details>` block for the sake of the thread's readability:

<details>
  <summary>
Stack trace
</summary>

```bash
pnpm build

> ~TODO~@0.0.1 build /Users/nickmazuk/Documents/Github/sveltekit-sharp
> svelte-kit build

vite v2.3.3 building for production...
✓ 14 modules transformed.
.svelte-kit/output/client/_app/manifest.json                    1.21kb
.svelte-kit/output/client/_app/layout.svelte-39a4acb4.js        0.49kb / brotli: 0.29kb
.svelte-kit/output/client/_app/pages/index.svelte-af73b2f3.js   0.79kb / brotli: 0.42kb
.svelte-kit/output/client/_app/error.svelte-8eafbf20.js         1.19kb / brotli: 0.57kb
.svelte-kit/output/client/_app/assets/start-a8cd1609.css        0.16kb / brotli: 0.10kb
.svelte-kit/output/client/_app/chunks/vendor-3681a316.js        5.05kb / brotli: 1.98kb
.svelte-kit/output/client/_app/start-22b096ce.js                17.18kb / brotli: 5.50kb
vite v2.3.3 building SSR bundle for production...
transforming (1) .svelte-kit/build/app.js"default" is imported from external module "sharp" but never used in "src/routes/endpoint.js".
✓ 10 modules transformed.
.svelte-kit/output/server/app.js   9.40kb

Run npm run preview to preview your production build locally.

> Using @sveltejs/adapter-vercel
 > node_modules/.pnpm/sharp@0.28.3/node_modules/sharp/lib/utility.js:7:22: error: No loader is configured for ".node" files: node_modules/.pnpm/sharp@0.28.3/node_modules/sharp/build/Release/sharp.node
    7 │ const sharp = require('../build/Release/sharp.node');
      ╵                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Build failed with 1 error:
node_modules/.pnpm/sharp@0.28.3/node_modules/sharp/lib/utility.js:7:22: error: No loader is configured for ".node" files: node_modules/.pnpm/sharp@0.28.3/node_modules/sharp/build/Release/sharp.node
Error: Build failed with 1 error:
node_modules/.pnpm/sharp@0.28.3/node_modules/sharp/lib/utility.js:7:22: error: No loader is configured for ".node" files: node_modules/.pnpm/sharp@0.28.3/node_modules/sharp/build/Release/sharp.node
    at failureErrorWithLog (/Users/nickmazuk/Documents/Github/sveltekit-sharp/node_modules/.pnpm/esbuild@0.11.23/node_modules/esbuild/lib/main.js:1443:15)
    at /Users/nickmazuk/Documents/Github/sveltekit-sharp/node_modules/.pnpm/esbuild@0.11.23/node_modules/esbuild/lib/main.js:1125:28
    at runOnEndCallbacks (/Users/nickmazuk/Documents/Github/sveltekit-sharp/node_modules/.pnpm/esbuild@0.11.23/node_modules/esbuild/lib/main.js:1043:65)
    at buildResponseToResult (/Users/nickmazuk/Documents/Github/sveltekit-sharp/node_modules/.pnpm/esbuild@0.11.23/node_modules/esbuild/lib/main.js:1123:7)
    at /Users/nickmazuk/Documents/Github/sveltekit-sharp/node_modules/.pnpm/esbuild@0.11.23/node_modules/esbuild/lib/main.js:1230:14
    at /Users/nickmazuk/Documents/Github/sveltekit-sharp/node_modules/.pnpm/esbuild@0.11.23/node_modules/esbuild/lib/main.js:606:9
    at handleIncomingPacket (/Users/nickmazuk/Documents/Github/sveltekit-sharp/node_modules/.pnpm/esbuild@0.11.23/node_modules/esbuild/lib/main.js:703:9)
    at Socket.readFromStdout (/Users/nickmazuk/Documents/Github/sveltekit-sharp/node_modules/.pnpm/esbuild@0.11.23/node_modules/esbuild/lib/main.js:573:7)
    at Socket.emit (node:events:365:28)
    at addChunk (node:internal/streams/readable:314:12)
 ERROR  Command failed with exit code 1.
```

</details>

**Information about your SvelteKit Installation:**

<details>
  <summary>Diagnostics</summary>

System:
    OS: macOS 11.3
    CPU: (16) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 8.95 GB / 64.00 GB
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 16.1.0 - ~/.nvm/versions/node/v16.1.0/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 7.11.2 - ~/.nvm/versions/node/v16.1.0/bin/npm
  Browsers:
    Brave Browser: 90.1.24.86
    Firefox: 87.0
    Safari: 14.1
  npmPackages:
    @sveltejs/kit: next => 1.0.0-next.109 
    svelte: ^3.34.0 => 3.38.2

- Adapter: `"@sveltejs/adapter-vercel": "^1.0.0-next.19"`
- Sharp: `"sharp": "^0.28.3"`
</details>

**Severity**

Blocking deployment of new feature.

**Comment by benmccann** [comment]
Since I wasn't aware of them, `.node` files are the extension for native addons: https://nodejs.org/api/addons.html

Does this work with `adapter-node`? Does Vercel support `.node` files independent of SvelteKit?

**Comment by Nick-Mazuk** [comment]
Building works with `adapter-node`, and I'd assume Vercel supports `.node` files since [Next.js uses the same `sharp` library](https://nextjs.org/docs/messages/install-sharp). It's also worth mentioning that I also run into issues building with the Vercel adapter locally, so it's not isolated to building on their servers.

**Comment by benmccann** [comment]
I wonder if it has anything to do with `adapter-vercel`'s use of esbuild:

https://github.com/sveltejs/kit/blob/6372690df51257cb811185d33892b6b3146620e8/packages/adapter-vercel/index.js#L30

**Comment by Nick-Mazuk** [comment]
Looks like esbuild is probably the culprit: https://github.com/evanw/esbuild/issues/1051

Luckily, the issue also has a code snippet for an ESBuild plugin to fix this. Not sure if adding it to the adapter itself is the direction Sveltekit should go, but it would likely work.

Alternatively, `sharp` (and equivalent packages) can also be marked as external. I don't think it makes sense to embed a list of external packages like `sharp` into the adapter. Perhaps there's some way to configure this? Perhaps the adapter could also follow the config `kit.vit.optimizeDeps.exclude`?

**Comment by cupcakearmy** [comment]
Hitting this issue as well

**Comment by 7antra** [comment]
Oupsi, let's go for a selfmade linode deploy as always :'(

**Comment by alteredorange** [comment]
Is there a current workaround for this? I can't get `sharp` to work, even trying:
```
 vite: {
      optimizeDeps: {
     //   include: ['sharp']  - also tried include
        exclude: ['sharp']
      },
      ssr: {
        external: ['sharp']
      },
```
Any help would be appreciated!

**Comment by Nick-Mazuk** [comment]
@alteredorange, I couldn't figure out a workaround so I ended up creating a custom [serverless function on Vercel](https://vercel.com/docs/serverless-functions/introduction). It's very much not ideal, though.

**Comment by 7antra** [comment]
Did someone find a solution here ?

**Comment by haishanh** [comment]
Try add a `postbuild` inside `scripts` in package.json

In my case, I'm using Prisma with N-API I have:

```
  "postbuild": "cp node_modules/.prisma/client/*.node .vercel_build_output/functions/node/render/;cp prisma/schema.prisma .vercel_build_output/functions/node/render/;ls .vercel_build_output/functions/node/render/",
```

see also [this comment I posted in Prisma repo](https://github.com/prisma/prisma/issues/7019#issuecomment-848026460).

**Comment by Nick-Mazuk** [comment]
@haishanh, with Primsa, did you also have errors during the build step? It looks like the issue you commented on in the Prisma repo didn't experience build-time errors. If you did also get those build-time errors, how'd you fix that?

**Comment by Nick-Mazuk** [comment]
Looks like this issue might be resolved once https://github.com/sveltejs/kit/pull/1732 is merged.

**Comment by Critical94** [comment]
Same Issue with the netlify-adapter here.

**Comment by Nick-Mazuk** [comment]
I haven't tried it yet, but it looks like https://github.com/sveltejs/kit/pull/1914 is merged and should fix this issue for all the official adapters (except static)

**Comment by 7antra** [comment]
Did someone try to use `sharp` with SvelteKit since `~v.151` ? Because I'm not sure is it possible : [Issue #1051 in esbuild github](https://github.com/evanw/esbuild/issues/1051)

**Comment by hmaesta** [comment]
I am **_not_** using Vercel adapter but facing this problem as well.

I was running **[svimg](https://github.com/xiphux/svimg)** on a **[Codesandbox](https://codesandbox.io/s/angry-rgb-z2zhy?file=/svelte.config.js)** and it was fine until I updated some packages to try something. Now everything is broken 😖

Even if I set old packages versions in `package.json` it is still returning a `sharp` error 🤔

**Comment by antony** [comment]
I got the .node files to work by rewriting the path in the lambda and copying the .node file to the lambda dir:

```
#!/bin/bash

LAMBDA_DIR=.vercel_build_output/functions/node/render/
cp node_modules/sharp/build/Release/*.node $LAMBDA_DIR
sed -i 's#\.\./build/Release#\.#g' $LAMBDA_DIR/index.js
```

However, it looks like once deployed to Vercel's lambda, there is no `vips` library available.

This seems odd because nextjs uses it.

Perhaps @leo can shine some light here - how is next using libvips? Is it possible for us to access it at runtime?

**Comment by leo** [comment]
Am I correctly understanding that you're in need of a native dependency in the Vercel Build Step? If so, which one exactly and what would it be used for?

If you're talking about optimizing images, have you considered using the respective [edge network API](https://vercel.com/docs/file-system-api#configuration/images) for that?

**Comment by antony** [comment]
@leo what we'd want, as I understand it, is for Vercel's build step to handle unbundled native dependencies, or somehow provide a way to allow these dependencies to become available to projects.

Libraries like sharp dynamically load `node` files at runtime, so Vercel's bundling doesn't factor them in and they are missing after deployment.

The fix above makes a manual move of the required files into the lambda dir, which is quite painful to figure out.

Didn't know about the edge network API though - that could save me a lot of time in a few places!

**Comment by leo** [comment]
I see. Thanks for clarifying!

When using the [File System API](https://vercel.com/docs/file-system-api), Vercel won't do any automatic bundling based on resolving what a Serverless Function needs like we do for our user-facing API. Instead, as described on the documentation page, we just place the file we're provided with directly in the final Lambda, and place any files next to it that were put into the same folder next to that entrypoint file, so that's where you could place them.

Does that help?

When it comes to this specific case, however, I recommend using the Image Optimization API, as anything else, I think, would be a waste of performance both in the build step and at runtime.

**Comment by antony** [comment]
Hi Leo - Ok so I think this makes sense. If I move the files into the right place, everything seems to work fine. I think the new issue I'm encountering is to do with vite's handing of `.node` files, which is a different issue I need to look at!

Thanks for your help!
